### ISSUE TITLE
Android mediascanner endless loop

### ISSUE THREAD

**Issue by karliwalti** [part-cw/lambdanative#295]
The mediascanner settings in eventloop.scm are causing android apps not to close properly. This is caused by a routine introduced in #95 
https://github.com/part-cw/lambdanative/blob/b1693b6e0bfdcb60405872963de5b205fa03aedf/modules/eventloop/eventloop.scm#L138

default should be set to true?

https://github.com/part-cw/lambdanative/blob/b1693b6e0bfdcb60405872963de5b205fa03aedf/modules/eventloop/eventloop.scm#L140

This will never become true again...

https://github.com/part-cw/lambdanative/blob/b1693b6e0bfdcb60405872963de5b205fa03aedf/modules/eventloop/eventloop.scm#L264

This will never exit. Reopening the app will lead to a grey screen. 

These lines would optionally disable the mediascanner upon termination, but it would be nicer to have a responsive mediascanner variable
```
(define (terminate . noscanner)
  (if app:android? (begin
    (let ((nm (not (if (= (length noscanner) 1) (car noscanner) #f))))
     (if nm (android-run-mediascanner))
    (android-finish)
   (if nm (let loop ()
     (if (not android-mediascanner-done?) (begin (thread-sleep! 0.1) (loop))))))))
  (if (procedure? hook:terminate) (hook:terminate))
)
```

**Comment by karliwalti** [comment]
Small precision:
https://github.com/part-cw/lambdanative/blob/b1693b6e0bfdcb60405872963de5b205fa03aedf/modules/eventloop/eventloop.scm#L226

does have a callback, but that does not seem to happen.

**Comment by mgorges** [comment]
I am not quite following your logic, isn't line 227 the place where android-mediascanner-done? is set to #t?

**Comment by karliwalti** [comment]
Yes. It would. Does not seem to happen when mediascanner is initiated in 

https://github.com/part-cw/lambdanative/blob/b1693b6e0bfdcb60405872963de5b205fa03aedf/modules/eventloop/eventloop.scm#L261

I can't pin it down more precisely atm. Android apps hang, unless line 263f are commented.

I do get a callback when the function is called during normal operation.

**Comment by peterlew** [comment]
Hi @karliwalti,

Do you have an example project involving mediascanner I can use to reproduce this issue?

Thanks,
Peter

**Comment by karliwalti** [comment]
Any of the demo apps will show that behavior. 
I did build with android  21 and 22

**Comment by peterlew** [comment]
I built DemoCube for Android APIs 19, 21, 22, and 23. In each case, I was able to run the app fine on a device - suspending, resuming, quitting, and re-opening the app with no problems. Perhaps you could try DemoCube on your device, and let me know what device it is and what Android version it is running. 

To be sure I understand, you get a grey screen when re-opening an app after it has been closed?

**Comment by karliwalti** [comment]
I have now done a small video with DemoHelloWorld. API 24 and the latest lambdanative code. The back button closes the app (put into background) -> launch the app (=> only black screen appears) -> 2x in video -> kill the app -> launch the app => all normal
Commenting lines mentioned above  will create the expected behavior.  video: https://photos.app.goo.gl/woweCtLvgfZnX7fh9

**Comment by peterlew** [comment]
Thank you @karliwalti! I can now reproduce this on my phone, using API 24. I'll be looking into it. Interestingly, this only happens for me if the app is paused with the back button, not via the home button or apps button.

**Comment by karliwalti** [comment]
yeah, this is because you would not call a (terminate) with this actions immediately

**Comment by peterlew** [comment]
I've found that the difference between the back button and home button is that `isFinishing()` evaluates to true on this line:

https://github.com/part-cw/lambdanative/blob/b6b088f2c1cb016073a7a31c3572c52a881535ff/loaders/android/bootstrap.java.in#L179

which means the following line is not run

https://github.com/part-cw/lambdanative/blob/b6b088f2c1cb016073a7a31c3572c52a881535ff/loaders/android/bootstrap.java.in#L180

when the back button is pressed. I don't think there's much harm in calling `mGLView.onPause()` even if the app is finishing, so I'd propose taking the `isFinishing()` check out altogether. This fixes the bug and doesn't break HybridApps where the back button event is captured and handled differently.

**Comment by mgorges** [comment]
The way I read https://developer.android.com/reference/android/app/Activity#isFinishing() that mean that the back button is actually killing the app, while the home button only closes it temporarily? So if you removed the test the rendering would stop, which might impact some closing up activities, as we trigger the heartbeat from there?

**Comment by peterlew** [comment]
Ah, ok. I'll look for a better way to handle the event, or have android-mediascanner exit correctly

**Comment by mgorges** [comment]
this may still get called, all I am asking is to check if the fifth element of the main function, which closes things gets called in this scenario, before you remote line 179. thanks!

**Comment by peterlew** [comment]
Makes sense, will do - thank you!

**Comment by peterlew** [comment]
After some further investigation, it seems android apps are not handling termination correctly at all. I'm not sure how much it has to do with mediascanner though -- at least on my phone, code in [terminate](https://github.com/part-cw/lambdanative/blob/b1693b6e0bfdcb60405872963de5b205fa03aedf/modules/eventloop/eventloop.scm#L259) never gets called, no matter how the app is exited. I can comment out all the mediascanner lines and have the same problem.
